modules/distribution-zones/src/test/java/org/apache/ignite/internal/distributionzones/DistributionZoneAwaitDataNodesTest.java

```diff
@@ -0,0 +1,910 @@
+     */
+    @Test
+    void testSeveralScaleUpAndSeveralScaleDownThenScaleDown() throws Exception {
+        startZoneManager(0);
+
+        TestSeveralScaleUpAndSeveralScaleDownDataObject testData = testSeveralScaleUpAndSeveralScaleDownGeneral();
+
+        LOG.info("Topology with removed nodes.");
+
+        Set<String> dataNodes = Set.of("node0");
+
+        topologyWatchListenerOnUpdate(dataNodes, testData.topVer2, testData.dataNodesRevision2);
+
+        assertTrue(testData.topVerUpFut2.isDone());
+        assertTrue(testData.topVerDownFut2.isDone());
+
+        CompletableFuture<Void> revisionDownFut = distributionZoneManager.zonesState().get(DEFAULT_ZONE_ID)
+                .scaleDownRevisionTracker().waitFor((long) testData.dataNodesRevision2);
+
+        assertFalse(revisionDownFut.isDone());
+
+        dataNodesWatchListenerOnUpdate(DEFAULT_ZONE_ID, dataNodes, false, testData.dataNodesRevision2,
+                testData.dataNodesRevision2 + 1);
+
+        assertTrue(waitForCondition(revisionDownFut::isDone, 3000));
+        assertEquals(dataNodes, testData.dataNodesUpFut3.get(3, SECONDS));
+        assertEquals(dataNodes, testData.dataNodesDownFut3.get(3, SECONDS));
+    }
+
+    private static class TestSeveralScaleUpAndSeveralScaleDownDataObject {
+        private final long topVer2;
+        private final long dataNodesRevision2;
+        private final CompletableFuture<Void> topVerUpFut2;
+        private final CompletableFuture<Void> topVerDownFut2;
+        private final CompletableFuture<Set<String>> dataNodesUpFut3;
+        private final CompletableFuture<Set<String>> dataNodesDownFut3;
+
+        TestSeveralScaleUpAndSeveralScaleDownDataObject(
+                long topVer2,
+                long dataNodesRevision2,
+                CompletableFuture<Void> topVerUpFut2,
+                CompletableFuture<Void> topVerDownFut2,
+                CompletableFuture<Set<String>> dataNodesUpFut3,
+                CompletableFuture<Set<String>> dataNodesDownFut3) {
+            this.topVer2 = topVer2;
+            this.dataNodesRevision2 = dataNodesRevision2;
+            this.topVerUpFut2 = topVerUpFut2;
+            this.topVerDownFut2 = topVerDownFut2;
+            this.dataNodesUpFut3 = dataNodesUpFut3;
+            this.dataNodesDownFut3 = dataNodesDownFut3;
+        }
+    }
+
+    /**
+     * This method invokes {@link DistributionZoneManager#topologyVersionedDataNodes(int, long)} with default and non-default zone id
+     * and different logical topology version. Collects data nodes futures.
+     * Simulates new logical topology with new nodes. Check that some of data nodes futures are completed.
+     * Simulates new logical topology with removed nodes. Check that some of data nodes futures are completed.
+     *
+     * @return Structure with data for continue testing.
+     */
+    private TestSeveralScaleUpAndSeveralScaleDownDataObject testSeveralScaleUpAndSeveralScaleDownGeneral()
```

Comment:
I do not understand this concept with `TestSeveralScaleUpAndSeveralScaleDownDataObject`, many things are done in one method, it is hard to read this code, please do something with that.

Reply:
I've removed TestSeveralScaleUpAndSeveralScaleDownDataObject